tools/closed-by-commits

```diff
@@ -0,0 +1,434 @@
+        object(oid: $oid) {
+          ... on Commit {
+            messageBody
+            associatedPullRequests(first: 10) {
+              nodes {
+                number
+                title
+                url
+                body
+                comments(first: 100) {
+                  nodes {
+                    body
+                  }
+                }
+                closingIssuesReferences(first: 50) {
+                  nodes {
+                    number
+                    title
+                    url
+                    body
+                    comments(first: 100) {
+                      nodes {
+                        body
+                      }
+                    }
+                    timelineItems(first:100, itemTypes:MARKED_AS_DUPLICATE_EVENT) {
+                      ... on IssueTimelineItemsConnection {
+                        nodes {
+                          ... on MarkedAsDuplicateEvent {
+                            duplicate {
+                              ... on Issue {
+                                number
+                              }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+
+    ISSUE_QUERY = """
+    query($number: Int!, $repo: String!) {
+      repository(owner: "zulip", name: $repo) {
+        issue(number: $number) {
+          number
+          title
+          body
+          comments(first: 100) {
+            nodes {
+              body
+            }
+          }
+          timelineItems(first:100, itemTypes:MARKED_AS_DUPLICATE_EVENT) {
+            ... on IssueTimelineItemsConnection {
+              nodes {
+                ... on MarkedAsDuplicateEvent {
+                  duplicate {
+                    ... on Issue {
+                      number
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+
+    def __init__(self, token: str, reponame: str, czo_issue_prefix: str) -> None:
+        self.github = Github(auth=Auth.Token(token))
+        self.reponame = reponame
+        self.czo_issue_prefix = czo_issue_prefix
+
+    @staticmethod
+    def _extract_czo_urls(text: str | None) -> set[str]:
+        if not text:
+            return set()
+
+        matches = re.findall(r"https://chat\.zulip\.org/[^\s\)\]\>]+", text)
+
+        urls = set()
+        for url in matches:
+            parsed_url = re.match(r"(.*)/topic/([^/]+)(/(near|with)/.*)?$", url)
+            if not parsed_url:
+                continue
+            # Normalize the topic by decoding and re-encoding
+            urls.add(
+                parsed_url[1]
+                + "/topic/"
+                + encode_hash_component(urllib.parse.unquote(parsed_url[2].replace(".", "%")))
+            )
```

Comment:
It's a bit subtle in looking at this version to see that the /with/  or /near/ gets dropped. (Which also leaves the reader potentially wondering if that's unintentional.)

It'd be good to have a brief comment mentioning that, and the rationale (https://github.com/zulip/zulip/pull/36288#discussion_r2445178732).